pkg/controllers/cloud/controller_sync_status.go

```diff
@@ -0,0 +1,141 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	antreav1alpha2 "antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	crdv1alpha1 "antrea.io/nephe/apis/crd/v1alpha1"
+	"antrea.io/nephe/pkg/logging"
+)
+
+var (
+	// map to track sync status of controllers.
+	controllerStatusMap map[string]bool
+
+	log = func() logging.Logger {
+		return logging.GetLogger("controllers-sync-status")
+	}
+)
+
+func init() {
+	controllerStatusMap = make(map[string]bool, 4)
+	// Add default entries into map, so that npController can check the
+	// sync status of all the controllers.
+	controllerStatusMap[CPAController] = false
+	controllerStatusMap[CESController] = false
+	controllerStatusMap[EEController] = false
+	controllerStatusMap[VMController] = false
+}
+
+// SetControllerSyncStatus sets the sync status of the controller.
+func SetControllerSyncStatus(controller string) {
+	// Lock is strictly not needed, as individual controller
+	// will be accessing a specific entry.
+	controllerStatusMap[controller] = true
+	log().Info("Sync done", "controller", controller)
+}
+
+// ResetControllerSyncStatus resets the sync status of the controller.
+func ResetControllerSyncStatus(controller string) {
+	controllerStatusMap[controller] = false
+}
+
+// IsControllerSynced returns true only when the controller is synced.
+func IsControllerSynced(controller string) bool {
+	return controllerStatusMap[controller]
+}
+
+// areControllersSynced returns true only when all the controllers are synced.
+func areControllersSynced(controllers []string) (bool, error) {
+	for _, name := range controllers {
+		if !controllerStatusMap[name] {
+			return false, fmt.Errorf("%v controller is syncing", name)
+		}
+	}
+	return true, nil
+}
+
+// waitForControllersToSync checks status of the controllers with a retry mechanism.
+func waitForControllersToSync(controllers []string, duration time.Duration) error { // nolint:unparam
+	operation := func() error {
+		_, err := areControllersSynced(controllers)
+		return err
+	}
+
+	b := backoff.NewExponentialBackOff()
+	b.MaxElapsedTime = duration
+	err := backoff.Retry(operation, b)
+	return err
+}
+
+// updatePendingSync decreases the pendingSyncCount.
+// If the count is 0, sets the controller sync status as done.
+func updatePendingSync(pendingSyncCount *int, controller string) {
+	if *pendingSyncCount > 0 {
+		*pendingSyncCount--
+		if *pendingSyncCount == 0 {
+			SetControllerSyncStatus(controller)
+		}
+	}
+}
+
+// waitTillControllerInitialized blocks until controller is initialized or timeout is reached.
+func waitTillControllerInitialized(initialized *bool, timeout time.Duration, controller string) error {
+	if err := wait.PollImmediate(time.Second*5, timeout, func() (bool, error) {
+		if *initialized {
+			log().Info("Initialize done", "controller", controller)
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		return fmt.Errorf("%v controller is initializing, err: %v", controller, err)
```

Comment:
controller initialization failed..

Reply:
Changed it to "failed to initialize %v controller".